docs/proposals/HealthChecks-L4.md

```diff
@@ -0,0 +1,287 @@
+# L4 HealthChecks
+
+## Context
+
+* at the moment, default configuration for dataplanes doesn't include health checks
+* without health checks (or, alternatively, without health statuses delivered via `EDS`), Envoy cannot exclude a non-responsive upstream endpoint from the list of valid targets for load balancing
+* there are several ways how Envoy can be configured to obtain health statuses for upstream endpoints
+  * Envoy can do "active" health checks
+    * for TCP it means that Envoy will periodically try establishing a connection
+  * Envoy can do "passive" health checks
+    * for TCP it means that Envoy will be remembering statuses of recent connection attempts
+  * Envoy can receive health statuses from the Control Plane as part of EDS API
+
+## Requirements
+
+1. support `Envoy -> upstream` "health checks"
```

Comment:
I don't think this scales very well. At least for active health checks.
Let's say we've got app `backend` and 10 other apps with 10 instances each that communicate with it. HC is sent every second. Now we generate 100rps to `backend`.

Reply:
It's always a user's choice.

Active health checks make perfect sense when your infrastructure is not big.